Secure Expense Dashboard with Express and Okta Blog (#1584)
oktadeveloper/okta-blog: _source/_posts/2025-07-07-express-authentication.md

---
layout: blog_post
title: "How to Build a Secure Expense Dashboard with Express and Okta"
author: akanksha-bhasin 
by: advocate

> communities: [security,javascript,go,python,mobile,java]

Reviewer: ```suggestion
communities: [javascript]
```
let's remove go, python, mobile, java, and security

description: "Build a secure expense dashboard using Express, Okta, and Passport with custom claims"

> tags: [okta, express, node, passport, oidc, oauth, javascript]

Reviewer: ```suggestion
tags: [express, node, passport, oidc, oauth, pkce, javascript]
```

might want to add pkce and remove okta (we'd already expect Okta here)

image: blog/express-okta-authentication/express-okta-authentication-social-image.jpeg

> type: awareness

Reviewer: ```suggestion
type: conversion
```

This is conversion as you're getting people to sign up for an account to try

> ---

Reviewer: We need to add the `github` frontmatter with the repo URL

Every web application needs authentication, but building it yourself is risky and time-consuming. Instead of starting from scratch, you can integrate Okta to manage user identity and use Passport with Express to simplify and secure the login flow. In this tutorial, you'll build a secure, passwordless, role-based expense dashboard where users can view their expenses tailored to their team.

Check out the complete source code on [GitHub](https://github.com/oktadev/okta-express-expense-dashboard-example) and get started without setting it up from scratch. 

**Table of Contents**{: .hide }
* Table of Contents
{:toc}

## Why use Okta for authentication 

Building an authentication system and handling credentials, sessions, and tokens is highly insecure and expose your application to serious vulnerabilities.

> Okta provides a secure, scalable, and standards-based solution using OpenID Connect (OIDC) and OAuth 2.0. It also integrates seamlessly with Node.js and Passport, supports passwordless login FIDO2 (WebAuthn), and allows you to customise ID tokens with team-specific claims. 

Reviewer: ```suggestion
Okta provides a secure, scalable, and standards-based solution using OpenID Connect (OIDC) and OAuth 2.0. It also integrates seamlessly with Express and Passport, supports passwordless login using FIDO2 and WebAuthn, and allows you to customise ID tokens with team-specific claims. 
```

Note - it's FIDO2 and WebAuthn combined (one is not a rename of the other)

> To further strengthen security, this project uses PKCE (Proof Key for Code Exchange), defined in [RFC 7636](https://www.rfc-editor.org/rfc/rfc7636). PKCE is a security extension to the Authorization Code flow. Originally designed for mobile apps, PKCE is now recommended for all OAuth clients, including web apps. It helps prevent CSRF and authorization code injection attacks and makes it useful for every type of OAuth client, even web apps that use client authentication. As the OAuth 2.0 and OIDC have steadily evolved over time, their security best practices have also advanced. [RFC 9700: Best Current Practice for OAuth 2.0 Security](https://www.rfc-editor.org/rfc/rfc9700.html) captures the current consensus on the most effective and secure implementation strategies.

Reviewer: ```suggestion
To further strengthen security, this project uses PKCE (Proof Key for Code Exchange), defined in [RFC 7636](https://www.rfc-editor.org/rfc/rfc7636). PKCE is a security extension to the Authorization Code flow. Originally designed for mobile apps, PKCE is now recommended for all OAuth clients, including web apps. It helps prevent CSRF and authorization code injection attacks and makes it useful for every type of OAuth client, even web apps that use client authentication. As OAuth 2.0 has steadily evolved over time, security best practices have also advanced. [RFC 9700: Best Current Practice for OAuth 2.0 Security](https://www.rfc-editor.org/rfc/rfc9700.html) captures the current consensus on the most effective and secure implementation strategies. The upcoming OAuth 2.1 draft codifies using PKCE for Authorization Code flow.
```

@Akankshabhasin please feel free to reword my suggestion. I think we should mention OAuth 2.1 in here and let people know this is going to be in the upcoming spec

> With Okta, you can easily implement modern authentication features and focus on your application logic without worrying about authentication infrastructure.

Reviewer: ```suggestion
With Okta, you can implement modern authentication features and focus on your application logic without worrying about authentication infrastructure.
```

Don't use words such as "easy", "simple", "just" as they can come across dismissive and exclusive. For some (many, including me at times), using Okta for their specific needs and project constraints isn't easy. Let's be inclusive in the way we position our posts.

> ##  Team-based expense dashboard using Express and Okta 

Reviewer: ```suggestion
##  Team-based expense dashboard using Express, Passport, OAuth 2.0, PKCE, and Okta 
```

Your original title doesn't contribute to SEO. People won't be searching for terms like team-based, expense (actually, they will search for "expense" but not for this reason 😂 ), and dashboard. 

Try to work in your keywords: Express, authentication and add other ones that reinforce, such as PKCE, OAuth, OIDC, RBAC, Authorization, etc

How about something like
> Team-based expense dashboard using Express, OAuth 2.0, OpenID Connect (OIDC), and RBAC

or

> Use OAuth 2.0, OpenID Connect (OIDC), and RBAC in an expense dashboard app

For any growing organization, tracking expenses by team isn't just helpful, it's essential. Let's build an expense dashboard where users log in with Okta and view spending data. Whether in Finance, Marketing or Support, each team sees a scoped view of expenses tied to their role. The users get clear visibility into what's being spent, by whom, and why.

Managers get transparency into their team's spending. At the same time, admins maintain a secure, centralized overview across all teams to track budgets, spending patterns, and ensure accountability, all without the headache of building authentication and security from scratch.

Along the way, you'll use OpenID Connect (OIDC), configure a custom authorization server, define custom claims for team-based access view, enforce role-based access directly from ID tokens, and integrate all the components into an Express application.

> ### Prerequisites 

Reviewer: ```suggestion
**Prerequisites**
```

Doesn't contribute to SEO. It doesn't need to be a header (with a TOC entry and anchor tag)

* Node.js installed (v22+ recommended)

* [Okta Integrator Free Plan org](https://developer.okta.com/signup/)

## Create your Express project and install dependencies

Open your terminal, create a new project folder, and install the necessary packages:

> `mkdir express-project-okta && cd express-project-okta`  

Reviewer: Why aren't we using codeblocks instead of inline?

also, I wouldn't include the command here. Instead, I'd say

> Create a new project folder named 'express-project-okta`. Open a terminal window in the project folder.

Or something to that effect. Your instructions won't work for Windows (outside of WSL)

Initialize a new Node.js project:

> `npm init -y`

Reviewer: ```suggestion
```sh
npm init -y
```
```

Why not use codeblocks with appropriate syntax highlighting??

Install the required packages:

> `npm install express passport openid-client express-session dotenv ejs express-ejs-layouts`

Reviewer: You need to pin versions (plus in codeblock) so it should be something like

```sh
npm install express@5.1 passport@blah .....
```

Having pinned versions protects the tutorial in case version changes cause breakage

Now, install the development dependencies:

`npm install --save-dev nodemon`

> **What do these dependencies do?** 

Reviewer: Nice section explaining the deps

These installed packages become your Express project's dependencies.

* **`express`**: Handles routing and HTTP middleware for your Node.js web app

* **`passport`**: Provides a flexible authentication framework

* **`openid-client`**: A server-side OpenID Relying Party implementation for Node.js runtime, including PKCE support

* **`express-session`**: Manages user sessions on the server

* **`dotenv`**: Loads environment variables from a `.env` file

* **`ejs`**: Enables dynamic HTML rendering using embedded JavaScript templates

* **`express-ejs-layouts`**: Adds layout support to EJS, helping manage common layout structures across views

> ## Create your .env file

Reviewer: ```suggestion
## Prepare the OIDC configuration required for authentication
```

Something like that? work in keywords.

Create a `.env` file in the root directory with placeholders for your Okta configuration.

> **Important Notes:** 
> * We will use a **Custom Authorization Server** to fetch custom claims from Okta.   
> * Write team names, e.g., Finance, Marketing, and Support. Make sure all are comma-separated, and if it's a two-word team name, you can add a space in between too. 

```
OKTA_ISSUER=https://{yourOktaDomain}/oauth2/default 
OKTA_CLIENT_ID={yourClientId}
OKTA_CLIENT_SECRET={clientSecret}
APP_BASE_URL=http://localhost:3000
POST_LOGOUT_URL=http://localhost:3000
ALL_TEAMS_NAME=TeamName1,TeamName2,Team Name3
```

You'll get these values from your Okta Admin Console in the next step.

## Create the Okta OIDC application

1. Sign up for a free [Integrator Free Plan](https://developer.okta.com/signup/). If you already have an account, [login](https://developer.okta.com/login/) to the [Okta Developer Console](https://developer.okta.com/signup/). 

> 2. Navigate to **Applications** \> **Create App Integration**.

Reviewer: ```suggestion
2. In the Okta Admin Console, navigate to **Applications** \> **Create App Integration**.
```

3. Choose:
   * **Sign-in method:** OIDC \- OpenID Connect

   * **Application type:** Web Application

4. Click **Next**.  
                             
5. Fill in:
   * **App integration name:** (e.g., `My Web App`)

   * **Sign-in redirect URIs:** `http://localhost:3000/authorization-code/callback`

   * **Sign-out redirect URIs:** `http://localhost:3000`

   * **Assignments:** Allow everyone in your organization to access.

6. After creating the app, click the edit button under Client Credentials and enable **Require PKCE as additional verification**.

7. Copy the **Client ID**, **Client Secret**, and **Okta Domain** and add them to your `.env` file.

> ## Set up passwordless login using FIDO2 with Okta

Reviewer: ```suggestion
## Set up passwordless login using FIDO2 and WebAuthn with Okta
```

This is an example of an excellent header. Clear and works in keywords

Add the **FIDO2 (WebAuthn)** authenticator in Okta to enable passwordless login and follow the [Okta documentation](https://help.okta.com/oie/en-us/content/topics/identity-engine/authenticators/configure-webauthn.htm) for complete enrollment and policy configuration instructions.

>  **Note:** When logging in for the first time, make sure you, as the admin, and all assigned users enroll in FIDO authentication. To simplify this process, you can create a user group containing all team members and assign it to the Web App. Then include the group in your enrollment policy, which you can customize as needed.

> ## Set up access policies in the Authorization Server

Reviewer: ```suggestion
## Set up access policies in the OAuth authorization server
```

You'll need to create an access policy and a corresponding rule in your Authorization Server. Refer to Okta's [documentation](https://developer.okta.com/docs/guides/configure-access-policy/main/) for complete instructions on configuring access policies.

> ## Set up custom claims for department information

Reviewer: Can you think of a stronger heading?

The most interesting part is to create and include custom attributes from the user profile, as Okta claims in your tokens. For example, you'll add the user's **`department`** information as a claim and display it as a team name on the expense dashboard.

> >  **Warning:** Avoid overloading ID tokens with too many custom claims. Keep ID tokens compact and efficient. Including unnecessary or excessive data can bloat the token and lead to network performance issues.

Reviewer: We need to make this louder and stronger. Specifically, any custom claim at all should require the decision whether the data is "required for user identity", and not for access control. If it doesn't need to be in the token, make as many API calls as required after authn. 

Perhaps even calling out that the userinfo endpoint has more data than what's default in the claim
